### ISSUE TITLE
realpath() open_basedir restriction in effect

### ISSUE THREAD

**Issue by andreasschroth** [getsentry/sentry-php#653]
New issue, but basically the same as reported in #512:

`realpath(): open_basedir restriction in effect. File(/usr/local/lib/php) is not within the allowed path(s)`
`"errfile":"/src/private/vendor/sentry/sentry/lib/Raven/Client.php","errline":319`

We don't use `open_basedir` anyway, as it was just a test balloon for a new config, but I noticed that this is a bit strange and unexpected for me. According to the other issue it should already be fixed, but for me, it looks like it's still open.

The docker container where I had it executed uses the very latest versions of everything - 1.9.2 of this lib, PHP 7.1.21, etc.

And keep up the good work here please, I like Sentry a lot. ;-)

**Comment by stayallive** [comment]
I'm not sure the fix mentioned there is possible 👍 

I tried the following:

```php
ini_set('open_basedir', __DIR__ . '/../../../');

@is_readable(__DIR__ . '/../../../../cannot/access/this');

var_export(error_get_last());die;
```

Still got:

```
array (
  'type' => 2,
  'message' => 'is_readable(): open_basedir restriction in effect. File(/private/var/www/sentry-php/test/Raven/Tests/../../../../cannot/access/this) is not within the allowed path(s): (/private/var/www/sentry-php/test/Raven/Tests/../../../)',
  'file' => '/private/var/www/sentry-php/test/Raven/Tests/ClientTest.php',
  'line' => 2697,
)```

So I don't think we can suppress this error with an `is_readable`... any clue what it is doing so it sees a file outside the `open_basedir`? Are you setting excluded paths or something?

**Comment by andreasschroth** [comment]
@stayallive I agree, afaik, `is_readable()` is not suited in this case at all. Because if you try to access / read from a path outside of `open_basedir` it immediately displays the warning. 

And yes, I also don't understand why the code executes a `realpath()` call on `'/usr/local/lib/php'`, that's obviously the folder where the `php` bin is inside.

**Comment by ste93cry** [comment]
I'm going to close the issue as it's quite old now and it seems that there is no proper solution to the problem. Feel free to reopen if it's still a problem

**Comment by andreasschroth** [comment]
@ste93cry For me, I don't understand why Sentry executes a `realpath()` on `/usr/local/lib/php`. Is there a specific reason for this? I'd provide a PR, but unfortunately, I'm not into the internals enough and couldn't figure out why the Sentry code executes a `realpath()` on the PHP binary.

**Comment by ste93cry** [comment]
The lines that generate the error according to what you posted are the following ones:

https://github.com/getsentry/sentry-php/blob/512d502979d3c7606f7091887fa659764c2cf480/lib/Raven/Client.php#L318-L319

In turn when the prefixes are set as configuration option the private function `_convertPath` is called and then a call to `realpath` is made. I think that the problem lies in the fact that the `include_path` includes the PHP executable which is outside the `open_basedir` path. I'm not sure about if or how we can solve the issue, though... The first thing I would take a look at is if this assumption is correct and if it's normal, but unfortunately I don't have any environment I can test this restriction on so a bit of debugging by someone else is needed to proceed

**Comment by andreasschroth** [comment]
@ste93cry Okay, thanks, I see. But the error message clearly says `realpath(): open_basedir restriction in effect.`. So probably the paths are then passed to `realpath()` somewhere in your code?

E.g. here: https://github.com/getsentry/sentry-php/blob/512d502979d3c7606f7091887fa659764c2cf480/lib/Raven/Client.php#L324

As said, unfortunately, I couldn't track down the exact error. I think it should be pretty easy though to create such an environment - just start a PHP docker container and change `open_basedir` to any folder that doesn't include the PHP bin.

**Comment by stayallive** [comment]
The only way to fix it is to read the openbasedir and make sure we never try and go outside that, but that seems silly.

They real question is how anything outside the realpath shows up in a stacktrace...

**Comment by ste93cry** [comment]
>So probably the paths are then passed to realpath() somewhere in the code?

Exactly, this is what I've written in the previous comment: I've explained all the flow that lead to this error with an assumption I've made about the setup that could trigger the error, but it's up to you to test it in your environment and tell us if it corresponds to the truth

>They real question is how anything outside the `realpath` shows up in a stacktrace...

This is a good question, we have no reason to read something outside the `open_basedir` path so imho it's something on your side wrongly configured

**Comment by ste93cry** [comment]
If you could at least share the `Dockerfile` or the exact configuration of PHP/Apache to replicate the problem it could be useful and would help us track down the issue. Otherwise, as it seems you're not interested anymore in the issue I will leave this issue closed and marked as "won't fix"

**Comment by andreasschroth** [comment]
> If you could at least share the `Dockerfile` or the exact configuration of PHP/Apache to replicate the problem it could be useful and would help us track down the issue. Otherwise, as it seems you're not interested anymore in the issue I will leave this issue closed and marked as "won't fix"

Well I checked the container from back then when I created the issue and `include_path` is set to `.:/usr/local/lib/php`. That's the default `include_path` from the official `php-fpm` docker container btw. - and `open_basedir = /project-root/`.

**Comment by dioncodes** [comment]
I had the same issue but was able to fix it by appending the /usr/share/php dir to the `open_basedir` variable (in addition to the web root, /tmp/ etc.).

**Comment by N-M** [comment]
I am encountering a similar error is there a solution for this without changing open_basedir settings?

```
Exception:
PHP Warning
Message:
realpath(): open_basedir restriction in effect. File(/opt/remi/php81/root/usr/share/pear) is not within the allowed path(s): (/var/www/)
Stacktrace:
realpath(/var/www/app.****.com/public_html/app-release-2024-09-16-14:36:18/backend/vendor/sentry/sentry/src/Options.php:1219)
Sentry\Options.normalizeAbsolutePath
array_map(/var/www/app.****.com/public_html/app-release-2024-09-16-14:36:18/backend/vendor/sentry/sentry/src/Options.php:1200)
Sentry\Options.Sentry\{closure}(/var/www/app.****.com/public_html/app-release-2024-09-16-14:36:18/backend/vendor/symfony/options-resolver/OptionsResolver.php:1128)
Symfony\Component\OptionsResolver\OptionsResolver.offsetGet(/var/www/app.****.com/public_html/app-release-2024-09-16-14:36:18/backend/vendor/symfony/options-resolver/OptionsResolver.php:905)
Symfony\Component\OptionsResolver\OptionsResolver.resolve(/var/www/app.****.com/public_html/app-release-2024-09-16-14:36:18/backend/vendor/sentry/sentry/src/Options.php:62)
Sentry\Options.__construct(/var/www/app.****.com/public_html/app-release-2024-09-16-14:36:18/backend/vendor/sentry/sentry/src/ClientBuilder.php:70)
Sentry\ClientBuilder.create(/var/www/app.****.com/public_html/app-release-2024-09-16-14:36:18/backend/src/Infrastructure/Factory/LoggerFactory.php:68)
App\Infrastructure\Factory\LoggerFactory.addHandler(/var/www/app.****.com/public_html/app-release-2024-09-16-14:36:18/backend/config/bootstrap.php:53)
require(/var/www/app.****.com/public_html/app-release-2024-09-16-14:36:18/public/app.php:10)
```

**Comment by stayallive** [comment]
Quick update for readers since this was discussed in Discord:

The `prefixes` options has a default value of `array_filter(explode(\PATH_SEPARATOR, get_include_path() ?: ''))`. This can cause issues if the include path includes paths not allowed by open basedir restrictions. Try overwriting the `prefixes` option to an array of paths that is allowed and makes sense for your project (these prefix paths are stripped from filenames to make them shorter in the UI).
